In [32]:
#imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import math
import random
from ISLP import load_data

Question 1

In [33]:
scores = [65, 70, 68, 73, 75, 80, 85, 78, 90, 100]

a.

In [34]:
mean = sum(scores) / len(scores)
mean

78.4

In [35]:
s = np.std(scores)
s

np.float64(10.229369482035537)

In [36]:
r = 2.262 * s / math.sqrt(len(scores))
print(mean - r)
print(mean + r)

71.08285828919516
85.71714171080485


The sample mean is 78.4, the standard deviation is 10.23, and the confidence interval is [71.08, 85.72].

b.

In [37]:
def bootstrap(data, n):
    means = []
    for j in range(n):
        indices = []
        bootstrap_sample = []
        for i in data:
            indices.append(random.randint(0,len(data)-1))
        for index in indices:
            bootstrap_sample.append(data[index])
        mean = sum(bootstrap_sample)/len(bootstrap_sample)
        means.append(mean)
    return means
    
means = bootstrap(scores, 40)
bootstrap_mean = sum(means) / len(means)
bootstrap_s = np.std(means)
bootstrap_r = 2.262 * bootstrap_s / math.sqrt(len(means))
print(f"The confidence interval is [{bootstrap_mean-bootstrap_r},{bootstrap_mean+bootstrap_r}]")

The confidence interval is [77.20061467352272,79.47438532647729]


c.

I would use the bootstrap method because the confidence interval that it gives, is much smaller, so we can be more sure of what population mean is.

Question 2.

a.

In [38]:
np.random.shuffle(scores)
subsets = np.array_split(scores, 5)
subsets

[array([73, 68]),
 array([70, 78]),
 array([90, 85]),
 array([80, 75]),
 array([100,  65])]

b.

In [39]:
subset_means = []
for subset in subsets:
    mean = 0
    for score in scores:
        if score not in subset:
            mean += score
    mean = mean / 8
    subset_means.append(mean)
subset_means

[80.375, 79.5, 76.125, 78.625, 77.375]

c.

In [40]:
maes = []
for i in range(0, len(subset_means)):
    mae = 0.5*abs(subset_means[i] - (subsets[0][0] + subsets[0][1])/2)
    maes.append(mae)
maes

[np.float64(4.9375),
 np.float64(4.5),
 np.float64(2.8125),
 np.float64(4.0625),
 np.float64(3.4375)]

d.

In [41]:
avg_mae = sum(maes) / len(maes)
print(f'The average mean absolute error is {avg_mae}')

The average mean absolute error is 3.95


e.

Questions 3

In [42]:
OJ = load_data('OJ')
OJ

,Purchase,WeekofPurchase,StoreID,PriceCH,PriceMM,DiscCH,DiscMM,SpecialCH,SpecialMM,LoyalCH,SalePriceMM,SalePriceCH,PriceDiff,Store7,PctDiscMM,PctDiscCH,ListPriceDiff,STORE
0,CH,237,1,1.75,1.99,0.00,0.00,0,0,0.500000,1.99,1.75,0.24,No,0.000000,0.000000,0.24,1
1,CH,239,1,1.75,1.99,0.00,0.30,0,1,0.600000,1.69,1.75,-0.06,No,0.150754,0.000000,0.24,1
2,CH,245,1,1.86,2.09,0.17,0.00,0,0,0.680000,2.09,1.69,0.40,No,0.000000,0.091398,0.23,1
3,MM,227,1,1.69,1.69,0.00,0.00,0,0,0.400000,1.69,1.69,0.00,No,0.000000,0.000000,0.00,1
4,CH,228,7,1.69,1.69,0.00,0.00,0,0,0.956535,1.69,1.69,0.00,Yes,0.000000,0.000000,0.00,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1065,CH,252,7,1.86,2.09,0.10,0.00,0,0,0.587822,2.09,1.76,0.33,Yes,0.000000,0.053763,0.23,0
1066,CH,256,7,1.86,2.18,0.00,0.00,0,0,0.670258,2.18,1.86,0.32,Yes,0.000000,0.000000,0.32,0
1067,MM,257,7,1.86,2.18,0.00,0.00,0,0,0.736206,2.18,1.86,0.32,Yes,0.000000,0.000000,0.32,0
1068,CH,261,7,1.86,2.13,0.00,0.24,0,0,0.588965,1.89,1.86,0.03,Yes,0.112676,0.000000,0.27,0
